packages/components/src/components/carousel/carousel.tsx

```diff
@@ -8,134 +8,211 @@ import {
+
+    const [selectedSnapIndex, setSelectedSnapIndex] = useState(0);
+
+    const [slidesInView, setSlidesInView] = useState<number[]>([0]);
+
+    const onSelect = useCallback((emblaApi: CarouselApi) => {
+      if (!emblaApi) {
+        return;
+      }
+
+      setSelectedSnapIndex(emblaApi.selectedScrollSnap());
+
+      setCanScrollPrev(emblaApi.canScrollPrev());
+      setCanScrollNext(emblaApi.canScrollNext());
+    }, []);
+
+    const scrollPrev = useCallback(() => {
+      api?.scrollPrev();
+    }, [api]);
+
+    const scrollNext = useCallback(() => {
+      api?.scrollNext();
+    }, [api]);
+
+    const handleKeyDown = useCallback(
+      (event: React.KeyboardEvent<HTMLDivElement>) => {
+        if (event.key === 'ArrowLeft') {
+          event.preventDefault();
+          scrollPrev();
+        } else if (event.key === 'ArrowRight') {
+          event.preventDefault();
+          scrollNext();
+        }
+      },
+      [scrollPrev, scrollNext],
+    );
+
+    useEffect(() => {
+      if (!api || !setApi) {
+        return;
+      }
+
+      setApi(api);
+    }, [api, setApi]);
+
+    useEffect(() => {
+      if (!api) {
+        return;
+      }
+
+      onSelect(api);
+      api.on('reInit', onSelect);
+      api.on('select', onSelect);
+      api.on('slidesInView', () => {
+        setSlidesInView(api.slidesInView());
+      });
+
+      return () => {
+        api.off('select', onSelect);
+      };
+    }, [api, onSelect]);
 
     return (
-      <CarouselContext.Provider value={[emblaRef, emblaApi]}>
-        <section
+      <CarouselContext.Provider
+        value={{
+          carouselRef,
+          api,
+          opts,
+          scrollPrev,
+          scrollNext,
+          canScrollPrev,
+          canScrollNext,
+          selectedSnapIndex,
+          slidesInView,
+        }}
+      >
+        <div
           aria-roledescription="carousel"
-          className={cn('relative -m-2 overflow-hidden p-2', className)}
+          className={cn('relative', className)}
+          onKeyDownCapture={handleKeyDown}
           ref={ref}
+          role="region"
           {...props}
         >
           {children}
-        </section>
+        </div>
       </CarouselContext.Provider>
     );
   },
 );
 
 Carousel.displayName = 'Carousel';
 
-type ForwardedRef = ElementRef<'div'> | null;
-
-const CarouselContent = forwardRef<ForwardedRef, ComponentPropsWithRef<'ul'>>(
+const CarouselContent = forwardRef<ElementRef<'div'>, ComponentPropsWithRef<'div'>>(
   ({ children, className, ...props }, ref) => {
-    const mutableRef = useRef<ForwardedRef>(null);
-    const [emblaRef] = useContext(CarouselContext);
-
-    useImperativeHandle<ForwardedRef, ForwardedRef>(ref, () => mutableRef.current, []);
-
-    const refCallback = useCallback(
-      (current: ForwardedRef) => {
-        emblaRef(current);
-        mutableRef.current = current;
-      },
-      [emblaRef],
-    );
```

Comment:
@matthewvolk Do you remember why we needed this?

Reply:
Yup! That was added based on an assumption that the ref was required for Makeswift registration. Now that we know the Content component won't be registered in Makeswift, we should definitely be safe to remove this. The `emblaRef` might still be required by Embla, but we should at least be able to remove the imperativeHandle and refCallback.